### PR TITLE
More Refactoring + Deep Sleep + Prevent WiFi Reconnect when reboot

### PR DESCRIPTION
**Pull request by timpur**
Main changes are:
- Small fix to String issue on config mode for older SDK versions (#429)
- Major Refactoring 
- Deep Sleep Function
- Prevent WiFi Reconnect when reboot

### CODE REVIEW

**Comment by timpur** [comment]
not sure if this is correct... its showing its going to pull previous commits that have already been merged?

File: src/Homie.cpp [hunk 1]
```diff
@@ -19,6 +19,7 @@ HomieClass::HomieClass()
   Interface::get().reset.triggerTime = DEFAULT_RESET_TIME;
   Interface::get().reset.resetFlag = false;
   Interface::get().flaggedForSleep = false;
+  Interface::get().flaggedForSleep = false;
```
Comment (luebbe) [inline src/Homie.cpp]: Any particular reason to do this twice?

Comment (timpur) [reply src/Homie.cpp]: It was an copy error. Fixed in the typo commit. Soz

Comment (luebbe) [reply src/Homie.cpp]: Just kidding, thx :)

**Comment by timpur** [comment]
@marvinroger Ready for review. Want to push this through before working on more enhancements. 
Going to also rebase my fork after this pull request.

Whats Next:
- Add incremental config updates via config mode (Web UI and API) (#433).
- Add ability to remove settings via incremental updates.
- Fix incremental update trying to add settings when previously no settings object exists.
- Better event buss handler for internal events and external. Hopefully this will make long async possesses that occur in homie more readable, like the prep for deep sleep procedure.
- Better Task Manger for single shot async tasks + ongoing loop async tasks that occur throughout homie. Might also expose this for external use for the user to register tasks and loops. It will be something like the [TickerScheduler](https://github.com/Toshik/TickerScheduler) but custom to fit homie's needs.
- More refactoring.
- Possibly Upload Firmware via config mode (Web UI and API).

**Comment by marvinroger** [comment]
@timpur thanks a lot! Can you please do a tiny commit to trigger a CircleCI build? I just reenabled it for PR. 😃

**Comment by timpur** [comment]
@marvinroger dont know what to do about the other checks ? Think these need fixing. Also not sure about lint....

**Comment by marvinroger** [comment]
I fixed the generate_docs step (I guess, it can only be tested on forks).

The linter is not happy because you do blocks that way:

```c++
if (1 + 1 == 2) {
}
else {
}
```

Whereas the Google style is:


```c++
if (1 + 1 == 2) {
} else {
}
```

Also, even though it might seem weird, we should not indent what's immediately inside a namespace "block". But don't worry, I'll take care of formatting the code. 👍 

Thanks!

**Comment by timpur** [comment]
@marvinroger no i also want to get the code style correct so we need to fix this in the `.editorconfig` so auto format follows these styles. I'll look into the google style c++.

Looks like `.editorconfig` might not do what we want. Thus ill look into other solutions.

**Comment by marvinroger** [comment]
@timpur editorconfig indeed only defines line endings, tabs/spaces, not much more. I am not sure there's a tool for that.

Just check out the changes I made: https://github.com/marvinroger/homie-esp8266/compare/e41bb9cd0aff069abe478379c4a109333dfdaaad...93a00e917b25c2a57d96b609fc0b2923981da7cd

**Comment by timpur** [comment]
@marvinroger looks good. What are you then using for auto formatting?

**Comment by marvinroger** [comment]
I don’t use any tools, I code that way. But that’d be an interesting
addition.
Clang might be suitable.

**Comment by timpur** [comment]
@marvinoger do you want to release beta 3 since this fixes an issue with beta 2.

#429

**Comment by bertmelis** [comment]
FYI, in VSCode and Atom you've got a cpplint extension.

**Comment by tripflex** [comment]
@bertmelis yeah that's what I use, VSCode with PlatformIO 😛

**Pull request merged**